File: src/assignment1_rt/src/distance_calc.cpp
 #include "rclcpp/rclcpp.hpp"
 #include "turtlesim/msg/pose.hpp"
 #include <iostream>
 #include "geometry_msgs/msg/twist.hpp"
 #include <cmath>
 #include "turtlesim/srv/spawn.hpp"
 #include <memory>
 #include "std_msgs/msg/string.hpp"

 using std::placeholders::_1;

  class Distance : public rclcpp::Node
 { 
    public:
    Distance(): Node("Distance_turtlesim")
     
    { 
        subscription_1 =this-> create_subscription<turtlesim::msg::Pose>("turtle1/pose",10,std::bind(&Distance::turtle1_callback, this, _1));
        subscription_2 =this-> create_subscription<turtlesim::msg::Pose>("turtle2/pose",10,std::bind(&Distance::turtle2_callback, this, _1));

        publisher_ = this->create_publisher<std_msgs::msg::String>("distance_topic",10);
        timer_ = this->create_wall_timer(std::chrono::milliseconds(50), std::bind(&Distance::timer_callback, this)); //wakes up every 0.5s
    }

 private:
    void turtle1_callback(const turtlesim::msg::Pose::SharedPtr msg)
    {
        t1_x= msg->x;
        t1_y= msg->y;
        t1_has_data = true;
    }

    void turtle2_callback(const turtlesim::msg::Pose::SharedPtr msg)
    {
        t2_x= msg->x;
        t2_y= msg->y;
        t2_has_data = true;
    }


    void timer_callback()
 {
    if (!t1_has_data || !t2_has_data) {
        return; 
    }
    
    raduis_=sqrt(pow(abs(t1_x - t2_x),2) + pow(abs(t1_y - t2_y),2));

    if ((t1_x > 9.5) || (t1_x < 0.5)){
        RCLCPP_WARN(this->get_logger(), "Turtle 1 out of bounds: x=%f", t1_x);
        distance_msg.data= "Turtle 1 out of bounds: x=" + std::to_string(t1_x);
        publisher_->publish(distance_msg);
    }
    else if ((t2_x >9.5) || (t2_x < 0.5)){
        RCLCPP_WARN(this->get_logger(), "Turtle 2 out of bounds: x=%f", t2_x);
        distance_msg.data= "Turtle 2 out of bounds: x=" + std::to_string(t2_x);
        publisher_->publish(distance_msg);
    }
    else if ((t2_y >9.5) || (t2_y <0.5)){               
        RCLCPP_WARN(this->get_logger(), "Turtle 2 out of bounds: y=%f", t2_y);
        distance_msg.data= "Turtle 2 out of bounds: y= " + std::to_string(t2_y);
        publisher_->publish(distance_msg);
    }
    else if ((t1_y>9.5) || (t1_y <0.5)){
        RCLCPP_WARN(this->get_logger(), "Turtle 1 out of bounds: y=%f", t1_y);
        distance_msg.data= "Turtle 1 out of bounds: y=" + std::to_string(t1_y);
        publisher_->publish(distance_msg);
    }else if (raduis_<=threshold_){
        distance_msg.data= "The turtles are too close! Their distance is:" + std::to_string(raduis_);
        RCLCPP_INFO(this->get_logger(), "Publishing distance: '%s'", distance_msg.data.c_str());
        publisher_->publish(distance_msg);
    }else {
        distance_msg.data= "The turtles are safe";
        RCLCPP_INFO(this->get_logger(), "Safe: '%s'", distance_msg.data.c_str());
        publisher_->publish(distance_msg);
        
    }
    

}

    std_msgs::msg::String distance_msg;
    geometry_msgs::msg::Twist twist_msg;
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr subscription_1;
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr subscription_2;
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
    
    
    //declaring variables
    double raduis_=0.0;
    double threshold_=2; //actuall threshold 1.5 gave a bit more for controls
    //defining turtle positions
    double t1_x = 5.54; 
    double t1_y = 5.54;
    double t2_x =7.0;
    double t2_y=7.0;

    //
    bool t2_has_data=false;
    bool t1_has_data=false;

 };

 int main(int argc, char * argv[])
 { 
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<Distance>());
    rclcpp::shutdown();
 
 return 0;

 }    

File: src/exam_example/exam_example/distance_calc.py
import rclpy
from rclpy.node import Node

from std_msgs.msg import String, Float64MultiArray
from turtlesim.msg import Pose
from geometry_msgs.msg import Twist, PoseArray
import math

class DistanceCalc(Node):

    def __init__(self):

        #have to intialize varibles
        # --- 1. DECLARING VARIABLES (Primitives) ---
        # In Python, we initialize them immediately.
        # Note: 'double' becomes just a float number.
        self.radius_ = 0.0
        self.threshold_ = 2.0  # actual threshold 1.5 gave a bit more for controls
    
        # Defining turtle positions
        self.t1_x = 5.54
        self.t1_y = 5.54
        self.t2_x = 7.0
        self.t2_y = 7.0

        # Booleans
        self.t2_has_data = False
        self.t1_has_data = False

        # --- 2. DECLARING MESSAGES ---
        # Instead of just declaring, we create the empty object instance here.
        self.distance_msg = String()
        self.twist_msg = Twist()
        self.obstacle_msg = String()
        self.mines_msg = String()

        self.obstacle_closest = 0.0 

        super().__init__('distance_calc') #name of node
        self.subscription_1 = self.create_subscription(Pose,"turtle1/pose",self.tutrle1_callback,10)
        self.subscription_2 = self.create_subscription(Pose,"turtle2/pose",self.tutrle2_callback,10)
        
        self.subscribe_mines = self.create_subscription(PoseArray, 'mines_topic',self.mine_callback, 10)
        self.publish_mines = self.create_publisher(String,'mine_detect',10)
        
        self.publisher_ = self.create_publisher(String, 'distance_topic', 10)

        self.subscriber_obs = self.create_subscription(Float64MultiArray, 'obstacle_topic',self.obstacle_callback, 10)
        self.publisher_obs = self.create_publisher(String,'obstacle_detect', 10)

        self.timer_ = self.create_timer(0.05, self.timer_callback)

    def tutrle1_callback(self, msg):
        self.t1_x = msg.x
        self.t1_y = msg.y
        self.t1_has_data = True

    def tutrle2_callback(self, msg):
        self.t2_x = msg.x
        self.t2_y = msg.y
        self.t2_has_data = True
        
    def obstacle_callback (self,msg)
        
        if len(msg.data) == 0
         return
        
        self.obstacle_closest = min(msg.data)

        if self.obstacle_closest < self.threshold_
            self.obstacle_msg = f"The obstacles are too close"
            self.get_logger().warn(f"Not safe:'{self.obstacle_msg.data}'")
        else
            self.obstacle_msg = f"safe"

    def mine_callback(self, msg):
         
        dist=[]

        # Check if "out of bounds" OR "too close" is inside the distance string
        # Reversed order logic compared to C++
        #its an array of poses so we need to loop through them
        for i in msg.poses:
            d = math.sqrt((i.position.x- self.t1_x)**2 + (i.position.y - self.t1_y)**2)
            dist.append(d)

        if min(dist) < self.threshold_:
            self.mines_msg.data= f"The mines are too close"
            self.get_logger().warn(f"Not safe:'{self.mines_msg.data}'")
        else:
            self.mines_msg.data= f"safe" 

        

    def timer_callback(self):
        
        
        self.publisher_obs.publish(self.obstacle_msg)

        self.publish_mines.publish(self.mines_msg)
        
        if not self.t1_has_data or not self.t2_has_data:
            return

        self.radius_ = math.sqrt(abs(self.t1_x - self.t2_x)**2 + abs(self.t2_y - self.t1_y)**2)

        if self.t1_x > 9.5 or self.t1_x < 0.5:
            self.get_logger().warn("Turtle 1 is out of bounds! x={self.t1_x}")
            self.distance_msg.data= f"Turtle 1 is out of bounds! x={self.t1_x}"
        elif self.t1_y > 9.5 or self.t1_y < 0.5:
            self.get_logger().warn("Turtle 1 is out of bounds! y={self.t1_y}")
            self.distance_msg.data= f"Turtle 1 is out of bounds! y={self.t1_y}"
        elif self.t2_x > 9.5 or self.t2_x < 0.5:
            self.get_logger().warn("Turtle 2 is out of bounds! x={self.t2_x}")
            self.distance_msg.data= f"Turtle 2 is out of bounds! x={self.t2_x}"
        elif self.t2_y > 9.5 or self.t2_y < 0.5:
            self.get_logger().warn("Turtle 2 is out of bounds! y={self.t2_y}")
            self.distance_msg.data= f"Turtle 2 is out of bounds! y={self.t2_y}"
        elif self.radius_ <= self.threshold_:
            self.distance_msg.data= f"The turtles are too close! Their distance is: {self.radius_}"
            self.get_logger().info(f"Publishing distance: '{self.distance_msg.data}'")
        else:
            self.distance_msg.data= f"The turtles are safe"
            self.get_logger().info(f"Safe:'{self.distance_msg.data}'")
        
        self.publisher_.publish(self.distance_msg)

def main(args=None):
    rclpy.init(args=args)
    distance_calc = DistanceCalc()
    rclpy.spin(distance_calc)
    rclpy.shutdown()

if __name__ == '__main__':
    main()
            
        
